LANGUAGE.md

```diff
@@ -0,0 +1,155 @@
+# Powdr Assembly (Powdr-Asm)
+
+Powdr-Asm is used to define instruction sets and write programs that depend on
+control flow, with the aim of generating ZKPs about them.
+Powdr-Asm follows the execution trace table model shared by many proof systems.
+In fact, Powdr-Asm extends Polygon Hermez' [PIL and
+zkASM](https://zkevm.polygon.technology/docs/category/polynomial-identity-language).
+In such model, the execution is represented by a two-dimensional fixed-size
+table, where each column represents a piece of state (for example registers or
+memory) and each row represents a step in time for a given execution.
+The program is represented by a transition system that specifies how the state
+evolves in time in the form of constraints.
+
+As an example, let's apply that model conceptually to a program that computes
+the factorial of a number. Our state consists of two registers:
+
+- CNT - a counter
+- ACC - the answer accumulator
+
+The first row will represent the initial state of our program, that is, `PC =
+0`, `CNT = <input>`, `ACC = 1`.
+From then on, we compute the subsequent rows with the following pseudocode:
+
+1. START:
+2. Is CNT = 0?
+3. If `yes` go to `END`
+4. ACC := ACC * CNT
+5. CNT := CNT - 1
+6. Go to START
+7. END:
+8. ACC equals `factorial(input)`
+
+We can compute the following execution trace for the specific input 5:
+
+| CNT | ACC |
+|-----|-----|
+| 5   | 1   |
+| 4   | 5   |
+| 3   | 20  |
+| 2   | 60  |
+| 1   | 120 |
+| 0   | **120** |
+
+Note that the size of the table determines how long the trace can be.
+
+## ISA
+
+There are three constructs that can be used to define an instruction set:
+
+- Registers
+- Macros
+- Instructions
+
+### Registers
+
+Registers are basic state components that can store a single numbers element
+each. There are three types of registers:
+
+- Program Counter (PC). Used for control flow.
+- Assignment registers. Used by parametric instructions.
+- General registers. Used by program logic.
+
+The registers above can be declared in the following ways:
+
+```asm
+reg pc[@pc];
+reg X[<=];
+reg A;
+```
+
+In the snippet above, the `@pc` annotation means that `pc` is implicitly
+incremented by one for instructions that do not use it.  The `<=` annotation
+means `X` is an assignment register and can be used by parametric instructions.
+Finally, `A` is a general purpose register and can be used by program logic.
+
+### Instructions
+
+Instructions are predicates that define constraints that must hold when the
+instruction is invoked. The constraints may refer to registers on the execution
+trace line that it is invoked and/or the next line.
+
+```asm
+instr ADD X, Y -> Z { X + Y = Z }
+```
+
+The above instruction can be read as "If `Z` is the same as `X + Y`, then `ADD
+X Y` returns `Z`".
+As an analogy, this is similar to Prolog predicates.
+
```

Comment:
If an instruction has formal parameters, these must either be literals or assignment registers.